### PR TITLE
Specifying interfaces in network partitions

### PR DESCRIPTION
**Pull request by eboasson**
This PR adds an "interface" attribute to the NetworkPartition configuration element, allowing easier configuration of the subset of network interfaces on which data is expected by readers mapping to that network partition. Without these changes, one must specify the host's IP addresses corresponding to those interfaces, which is very impractical and error-prone.

The other commits are a bit of refactoring/cleaning up in the affected code.

Creating as a draft because I think it is useful as is for those who happen to be specifying unicast addresses in the `NetworkPartition`s today, but this PR really needs to also update the configuration guide and include some tests for these features. (The old code was working but without tests, the new code has changes that are substantial enough to not allow "grandfathering" it in.)

### CODE REVIEW

File: src/core/ddsi/include/dds/ddsi/ddsi_ownip.h
Comment (dpotman) [inline src/core/ddsi/include/dds/ddsi/ddsi_ownip.h:27]: Maybe add `nw` or `network` in the name?

File: src/core/ddsi/src/ddsi_raweth.c
Comment (dpotman) [inline src/core/ddsi/src/ddsi_raweth.c:327]: Nit-picking: `loc` and `interf` are not unused args any more

File: src/core/ddsi/src/ddsi_nwpart.c
Comment (dpotman) [inline src/core/ddsi/src/ddsi_nwpart.c:335]: Is whitespace already stripped out of `np->interface_names`? If not, `strcmp` on line 339 might fail because of this?

File: src/core/ddsi/include/dds/ddsi/ddsi_nwpart.h
Comment (dpotman) [inline src/core/ddsi/include/dds/ddsi/ddsi_nwpart.h:18]: `extern "C"` is missing

**Review by dpotman — COMMENTED** [review]
Looks good! Just a few minor remarks, and as you mentioned it would be nice to have at least a test case that checks some basics of specifying the address and interface of network partitions.

File: src/core/ddsi/src/ddsi_nwpart.c
Comment (eboasson) [reply src/core/ddsi/src/ddsi_nwpart.c:335]: I figured it would be acceptable to treat all white space in the names as significant and that there really was no reason to allow empty entries. I'm open to changing it, it just seems like a needless complication in the code.

**Comment by eboasson** [comment]
Still have to do the tests, but I think otherwise it is starting to look halfway decent.

File: src/core/ddsi/src/ddsi_nwpart.c
Comment (dpotman) [reply src/core/ddsi/src/ddsi_nwpart.c:335]: I was triggered by the call to `strspn` in `convert_network_partition_addresses_one` that uses `", \t"`, but accepting leading/trailing white space in a name sounds reasonable, so I'm fine with the current code.

**Comment by eboasson** [comment]
@dpotman if you'd be willing to give it another look now that I have added some tests (and fixed a few things that seemed missing previously). Clearly the "DO NOT MERGE" commit is supposed not to be merged because I'd much rather merge #1397 first.

Once #1397 is merged, this one has had another look at it and I've squashed some of the commits, this is ready for merging, I'd say.

**Review by dpotman — APPROVED** [review]
I've reviewed the fixes and changes, LGTM!

**Comment by eboasson** [comment]
Thanks @dpotman !

**Pull request merged**
